### PR TITLE
refactor(expr): make `NOW` a separate expression type

### LINKED ISSUE #9605: Function bind error seems incorrect
### Describe the bug

image

I gave 1 arguments to `now`, which should accept 0 arguments, but the error report said (2, 1);

### To Reproduce

_No response_

### Expected behavior

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by BugenZhao**
Signed-off-by: Bugen Zhao <i@bugenzhao.com>I hereby agree to the terms of the [RisingWave Labs, Inc. Contributor License Agreement](https://gist.github.com/TennyZhuang/f00be7f16996ea48effb049aa7be4d66#file-rw_cla).

## What's changed and what's your intention?

`NOW()` has different semantics in different contexts:
- batch: a literal constant of the current time
- streaming: a retractable stream of the current time

In both cases, there's nothing called "`NOW` function" in the expression execution framework. To make it clear, this PR makes `NOW` a separate expression type which only appears in the frontend.

There's no breaking change as `NOW` did not appear in the persisted streaming plan originally.

Also close #9605.

## Checklist For Contributors

- [x] I have written necessary rustdoc comments
- [x] I have added necessary unit tests and integration tests
- [x] I have added fuzzing tests or opened an issue to track them. (Optional, recommended for new SQL features #7934).
- [x] I have demonstrated that backward compatibility is not broken by breaking changes and created issues to track deprecated features to be removed in the future. (Please refer to the issue)
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)

## Checklist For Reviewers

- [ ] I have requested macro/micro-benchmarks as this PR can affect performance substantially, and the results are shown.

## Documentation

- [x] My PR **DOES NOT** contain user-facing changes.

<details><summary>Click here for Documentation</summary>

### Types of user-facing changes

Please keep the types that apply to your changes, and remove the others.

- Installation and deployment
- Connector (sources & sinks)
- SQL commands, functions, and operators
- RisingWave cluster configuration changes
- Other (please specify in the release note below)

### Release note

</details>

### CODE REVIEW

**Comment by wangrunji0408** [comment]
Seems conflict with #9610.

**Comment by BugenZhao** [comment]
> Seems conflict with #9610. Could you help cherry-pick this patch?

After this PR, we don't have a `NOW` function anymore. 😄 And the arity check will be done in binder with `guard_by_len`.

File: src/frontend/src/binder/expr/function.rs [hunk 1]
```diff
@@ -343,17 +343,22 @@ impl Binder {
...
-                    )));
-                }
-                raw_call(ExprType::Now)(binder, inputs)
-            })
+            guard_by_len(
+                0,
+                raw(move |binder, _inputs| {
+                    binder.ensure_now_function_allowed()?;
+                    // `now()` in batch query will be convert to the binder time.
+                    if binder.is_for_batch() {
```
Comment (st1page) [inline src/frontend/src/binder/expr/function.rs:351]: some offline discussion: we can remove this and use ExprImpl::Now for batch to forbid split batch and stream logic in the binder.

File: src/frontend/src/expr/now.rs [hunk 2]
```diff
@@ -13,12 +13,25 @@
...
+impl Expr for Now {
+    fn return_type(&self) -> DataType {
+        DataType::Timestamptz
+    }
+
+    fn to_expr_proto(&self) -> ExprNode {
+        unreachable!(
+            "`Now` should be translated to `Literal` in batch mode or `NowNode` in stream mode"
+        )
+    }
```
Comment (st1page) [inline src/frontend/src/expr/now.rs:36]: It should be ensured by `count_nows`. In other words, I think before converting to proto, there should be a visitor to do a final check to return an error.

**Review by st1page — APPROVED** [review]
LGTM

**Review by soundOfDestiny — APPROVED** [review]
LGTM. thanks!

**Comment by soundOfDestiny** [comment]
> LGTM

You're too fast. I've just finished my review 😢

**Comment by wangrunji0408** [comment]
> > Seems conflict with #9610. Could you help cherry-pick this patch?
> 
> After this PR, we don't have a `NOW` function anymore. 😄 And the arity check will be done in binder with `guard_by_len`.

Let me close my PR. 😄

**Pull request merged**